MaxText/layers/attention_op.py

```diff
@@ -1210,19 +1223,42 @@ def compute_local_attention(
           local_max is the local max of exponentials
           local_sum is the sum of exponentials for this chunk, divided by exp(local_max).
     """
-    local_max = jnp.max(attn_weights, axis=-1, keepdims=True)
-    local_exps = jnp.exp(attn_weights - local_max)
-    local_sum = jnp.sum(local_exps, axis=-1, keepdims=True)
+    if sinks is not None:
+      # reshape attn weights to combine key/value and query group dimensions
+      b, n_kv, g, t, s = attn_weights.shape
+      n_q = n_kv * g
+      attn_weights_reshaped = jnp.reshape(attn_weights, (b, n_q, t, s))
+
+      # broadcast sinks to match the attn weights dimension and combine
+      sinks_param = sinks.astype(attn_weights.dtype)  # (n_q,)
+      sinks_logits = sinks_param[jnp.newaxis, :, jnp.newaxis, jnp.newaxis]  # (1, n_q, 1, 1)
+      sinks_logits = jnp.broadcast_to(sinks_logits, (b, n_q, t, 1))
+      combined_logits = jnp.concatenate([attn_weights_reshaped, sinks_logits], axis=-1)
+
+      # softmax
+      local_max = jnp.max(combined_logits, axis=-1, keepdims=True)
+      local_exps_combined = jnp.exp(combined_logits - local_max)
+      local_sum = jnp.sum(local_exps_combined, axis=-1, keepdims=True)
+
+      # reshape and transpose
+      local_exps = local_exps_combined[..., :s]
+      local_exps = jnp.reshape(local_exps, (b, n_kv, g, t, s))
+      local_max = jnp.transpose(local_max, (0, 2, 1, 3))  # (b, t, n_q, 1)
+      local_sum = jnp.transpose(local_sum, (0, 2, 1, 3))  # (b, t, n_q, 1)
+    else:
+      local_max = jnp.max(attn_weights, axis=-1, keepdims=True)
+      local_exps = jnp.exp(attn_weights - local_max)
+      local_sum = jnp.sum(local_exps, axis=-1, keepdims=True)
 
-    local_sum = jnp.moveaxis(local_sum, -2, 1)
-    local_max = jnp.moveaxis(local_max, -2, 1)
+      local_sum = jnp.moveaxis(local_sum, -2, 1)
+      local_max = jnp.moveaxis(local_max, -2, 1)
 
-    local_max = jnp.reshape(
-        local_max, (local_max.shape[0], local_max.shape[1], local_max.shape[2] * local_max.shape[3], 1)
-    )
-    local_sum = jnp.reshape(
-        local_sum, (local_sum.shape[0], local_sum.shape[1], local_sum.shape[2] * local_sum.shape[3], 1)
-    )
+      local_max = jnp.reshape(
+          local_max, (local_max.shape[0], local_max.shape[1], local_max.shape[2] * local_max.shape[3], 1)
+      )
+      local_sum = jnp.reshape(
+          local_sum, (local_sum.shape[0], local_sum.shape[1], local_sum.shape[2] * local_sum.shape[3], 1)
```

Comment:
is there any repeated code to reuse in this if/else?

Reply:
Sounds good, will combine as one.